File: app/src/main/java/ai/elimu/analytics/db/Converters.java
package ai.elimu.analytics.db;

import android.text.TextUtils;

import androidx.room.TypeConverter;

import java.util.Arrays;
import java.util.Calendar;

import ai.elimu.model.v2.enums.analytics.LearningEventType;

/**
 * See https://developer.android.com/training/data-storage/room/referencing-data
 */
public class Converters {

    @TypeConverter
    public static Calendar fromLong(Long value) {
        Calendar calendar = Calendar.getInstance();
        calendar.setTimeInMillis(value);
        return calendar;
    }

    @TypeConverter
    public static Long toLong(Calendar calendar) {
        return calendar.getTimeInMillis();
    }


    @TypeConverter
    public static LearningEventType fromString(String value) {
        LearningEventType learningEventType = null;
        if (!TextUtils.isEmpty(value)) {
            learningEventType = LearningEventType.valueOf(value);
        }
        return learningEventType;
    }

    @TypeConverter
    public static String toString(LearningEventType learningEventType) {
        String value = learningEventType.toString();
        return value;
    }


    @TypeConverter
    public static String[] fromArrayString(String value) {
        return value.substring(1, value.length() - 1).split(", ");
    }

    @TypeConverter
    public static String toString(String[] array) {
        return Arrays.toString(array);
    }
}

File: app/src/main/java/ai/elimu/analytics/db/Converters.kt
package ai.elimu.analytics.db

import ai.elimu.model.v2.enums.analytics.LearningEventType
import android.text.TextUtils
import androidx.room.TypeConverter
import java.util.Calendar

/**
 * See [Referencing complex data using Room](https://developer.android.com/training/data-storage/room/referencing-data)
 */
object Converters {
    @JvmStatic
    @TypeConverter
    fun fromLong(value: Long): Calendar {
        val calendar = Calendar.getInstance()
        calendar.setTimeInMillis(value)
        return calendar
    }

    @JvmStatic
    @TypeConverter
    fun toLong(calendar: Calendar): Long {
        return calendar.getTimeInMillis()
    }


    @JvmStatic
    @TypeConverter
    fun fromString(value: String?): LearningEventType? {
        var learningEventType: LearningEventType? = null
        if (!TextUtils.isEmpty(value)) {
            learningEventType = LearningEventType.valueOf(value!!)
        }
        return learningEventType
    }

    @JvmStatic
    @TypeConverter
    fun toString(learningEventType: LearningEventType?): String? {
        return learningEventType?.toString()
    }


    @JvmStatic
    @TypeConverter
    fun fromArrayString(value: String): Array<String?> {
        return value.substring(1, value.length - 1).split(", ".toRegex())
            .dropLastWhile { it.isEmpty() }.toTypedArray()
    }

    @TypeConverter
    fun toString(array: Array<String?>?): String {
        return array.contentToString()
    }
}
